bug fixes & updates

PoOnesNerfect opened the pull request:
Updates:
- fetch current terminals dark/light theme
- should queue instead of execute to avoid flushing multiple times
- only re-render changed lines to avoid flashing screen
- support hjkl navigation
- left/right/tab/backtab now jumps sections
- cargo.toml version field is a version requirement, not just a version (it was failing to parse like "=1.0.0")
- api should fetch crate by package name (name != package name)

Main issue I was facing were dark terminal theme not being supported, and screen flashing every time an action is taken.
I tried not to change the existing implementation as much as possible.

Also, ideally, we should look at the lockfile to fetch the actual versions of crates.

Please let me know what you think.

PoOnesNerfect committed:
bug fixes

- fetch current terminals dark/light theme
- should queue instead of execute to avoid flushing multiple times
- only re-render changed lines to avoid flashing screen
- support hjkl navigation
- left/right/tab/backtab now jumps sections
- cargo.toml version field is a version requirement, not just a semver
- api should fetch crate by package name

PoOnesNerfect committed:
make version strings bold

BenJeau commented:
Thanks for the change! At first look it seems overall good, will want to take a deeper look at the changes, probably over the next few days.

Seems like you caught a few bugs 😶, didn't know about queue and other things.

PoOnesNerfect committed:
fix bug

PoOnesNerfect commented:
> Thanks for the change! At first look it seems overall good, will want to take a deeper look at the changes, probably over the next few days.
> 
> Seems like you caught a few bugs 😶, didn't know about queue and other things.

Please take your time!

I'm not an expert in tui development, but wanted to contribute as it seems pretty handy for me :)

If I have time, I would like to also add some features like:
- looking at lockfile to fetch actual versions
- async loading of crate info so we don't have to wait for all apis to load, especially for large number of dependencies, it could take awhile
- better navigation features? like jumping a few lines at a time using shift-arrow
- scroll behavior?
- descriptions cut off because I use half a terminal screen using zellij, so maybe it'll be cool to scroll horizontally as well
- fuzzy search
- be able to fetch info from other registries that crates.io but not sure how feasible this is
- adding/removing crates would be pretty cool too

Let me know what you think!

BenJeau commented on src/api.rs:
```diff
@@ -33,7 +34,10 @@ fn get_field_from_versions(
         versions?
             .iter()
-            .find(|v| v.get("num").and_then(|v| v.as_str()).unwrap_or("") == version)?
+            .find(|v| {
+                v.get("num").and_then(|v| v.as_str()).unwrap_or("")
+                    == version.trim_start_matches(&['=', '^'])
+            })?
```
This function is for parsing the version from the Crates.io API. Have you encountered the field starting with `=` or `^`? Just curious, especially if you have an example API call URL.

If yes, could you add a little test at the bottom of this file for this new supported behaviour?

PoOnesNerfect replied:
the `version` field here for `current_version_date` is not from an API but from `CargoDependency.version` field which is actually `VersionReq` type.

So, when the version in `Cargo.toml` had the field value like `^0.1.0` it would not match any versions from API, so date would be `None`.

I have some updates locally that looks into lockfile to fetch versions instead of from `Cargo.toml`, so I'll push the commit here after cleaning it up a bit.

BenJeau commented on Cargo.toml:
```diff
@@ -11,11 +11,12 @@ categories = ["command-line-utilities"]
 [dependencies]
 clap = { version = "4.5.20", features = ["derive"] }
-clap-cargo = "0.14.1"
+clap-cargo = "0.15.1"
```
**Note to self:** After this PR, I'll create another MR updating the deps. I think a few of them have new version.

BenJeau commented on src/cli.rs:
```diff
@@ -149,30 +234,38 @@ impl State {
     }
 
-    fn render_dependencies(&mut self) -> Result<(), Box<dyn std::error::Error>> {
+    // parameter takes a specific indices to re-render
+    // if indices is empty, then re-render the entire list
```
Would be nice if these comments (and the ones below) were converted to doc comments with `///`

BenJeau commented on src/main.rs:
```diff
@@ -22,26 +28,29 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
+        outdated_deps,
+        total_deps,
+        args.all,
+        theme.join().expect("operation in thread failed"),
```
Wonder if we could default to a theme here if it fails?

PoOnesNerfect replied:
great idea!

BenJeau commented on src/cargo.rs:
```diff
@@ -33,14 +34,14 @@ impl PartialOrd for CargoDependency {
 impl CargoDependency {
     fn get_latest_version_wrapper(&self) -> Option<Dependency> {
-        let parsed_current_version = Version::parse(&self.version).ok()?;
+        let parsed_current_version_req = VersionReq::parse(&self.version).ok()?;
 
-        let response = api::get_latest_version(self).expect("Unable to reach crates.io");
+        let response = api::get_latest_version(self).expect("Unable to reach crates.io")?;
 
         let parsed_latest_version =
             Version::parse(&response.latest_version).expect("Latest version is not a valid semver");
 
-        if parsed_current_version < parsed_latest_version {
+        if !parsed_current_version_req.matches(&parsed_latest_version) {
```
This introduces a regression. It will evaluate if the current latest version satisfies the Cargo.toml dep version and not if it is an old version. I could see this being useful, but maybe behind a flag or we could rework it when/if we add logic with the lock file?

This changes the behaviour of the tool and will say that the deps are up to date while they might not be. E.g., if you run the tool on the Cargo.toml of this repo, it will show that 4 deps are out of date, but after this change, it will say that they are all up to date.

PoOnesNerfect replied:
good point, it was a limitation.

I have a local updates that uses `Cargo.lock` so I'll push the commit soon! It should fix some of these issues.

BenJeau commented on src/cli.rs:
```diff
@@ -53,48 +55,85 @@ impl Longest {
+                        self.selected = vec![false; self.outdated_deps.len()];
+                    } else {
+                        self.selected = vec![true; self.outdated_deps.len()];
+                    }
```
This could be simplified to something like:
```suggestion
                    let all_selected = self.selected.iter().all(|s| *s);
                    self.selected = vec![!all_selected; self.outdated_deps.len()];
```

BenJeau commented on src/cli.rs:
```diff
@@ -105,11 +144,67 @@ impl State {
+        } else if self.outdated_deps.0[0].kind == self.outdated_deps.0[self.cursor_location].kind {
+            self.cursor_location
+        } else {
+            0
+        }
+    }
+
+    fn prev_section_location(&self) -> usize {
+        let curr_kind = self.outdated_deps.0[self.cursor_location].kind;
+
+        let mut iter = self.outdated_deps.0[..self.cursor_location]
+            .iter()
+            .enumerate()
+            .rev()
+            .skip_while(|(_, d)| d.kind == curr_kind);
+
+        if let Some((i_, d_)) = iter.next() {
+            iter.take_while(|(_, d)| d.kind == d_.kind)
+                .last()
+                .map(|(i, _)| i)
+                .unwrap_or(i_)
+        } else if self.outdated_deps.0.last().unwrap().kind
+            == self.outdated_deps.0[self.cursor_location].kind
+        {
+            self.cursor_location
+        } else {
+            let last_kind = self.outdated_deps.0.last().unwrap().kind;
+
+            self.outdated_deps
+                .0
+                .iter()
+                .enumerate()
+                .rev()
+                .take_while(|(_, d)| d.kind == last_kind)
+                .last()
+                .unwrap()
+                .0
+        }
+    }
+
```
This could be simplified if we sort the dependencies based on the kinds and we assume they are sorted here. We could do something like this and replace where we called `prev_section_location()` with `change_section(false)` and `change_section(true)` for `next_section_location()`

```suggestion
    fn change_section(&mut self, next: bool) -> usize {
        let cursor_kind = self.outdated_deps.0[self.cursor_location].kind;
        let mut other_kind = None;
        let mut other_index = self.cursor_location;

        for i in 1..self.outdated_deps.len() {
            let index = if next {
                (self.cursor_location + i) % self.outdated_deps.len()
            } else {
                (self.cursor_location - i) % self.outdated_deps.len()
            };
            let curr_kind = self.outdated_deps.0[index].kind;

            if curr_kind != cursor_kind {
                if other_kind.is_none() {
                    other_kind = Some(curr_kind);
                    other_index = index;
                } else {
                    other_index = index;
                }
            }

            if other_kind.is_some() && (next || other_kind != Some(curr_kind)) {
                break;
            }
        }

        other_index
    }
```

Then we could make the following in the `new(...)` method of `Dependencies` as follows (or we could sort it somewhere else):

```rs
        let mut dependencies = dependencies;
        dependencies.sort_by(|a, b| a.kind.cmp(&b.kind));
        Self(dependencies)
 ```

Let me know what you think! Just think that this way we do not need to worry about some edge cases and can reuse it for both direction changing sections.

BenJeau replied:
After looking more in the codebase, sorting is already being done: https://github.com/BenJeau/cargo-interactive-update/blob/2571f9b8d0cadf04aa46bbf056fa7ca1d15d2ff2/src/cargo.rs#L68-L69
So no need to do the last part

PoOnesNerfect replied:
I like this! I updated the code as you suggested

BenJeau reviewed (commented):
Was a bit sick these past few days, so I'm just now getting around to looking at the changes.

Overall great work! I'm also no expert on TUIs, but I like the new section jumping, looking if package name is specified (forgot that you could rename a package in `Cargo.toml`...), and the new keybindings. Just have a few comments :)

As for your previous comment, as long as it doesn't break the current functionality in major ways, I'm all for it! If you want to discuss it more, we could create a new issue and discuss it there (to avoid clogging up this PR).

PoOnesNerfect committed:
Merge branch 'main' into bug-fixes

# Conflicts:
#	src/cargo.rs
#	src/cli.rs
#	src/dependency.rs
#	src/main.rs

PoOnesNerfect committed:
updates

- use lockfile to fetch crate versions
- rename `Dependency.package_name` to `Dependency.workspace_member`

PoOnesNerfect commented:
@BenJeau Addressed comments. PTAL!

I also have more PRs ready to open like being able to fetch from alternate private registries.

I'll actually create issues for them first.

PoOnesNerfect committed:
set theme to default if thread errors

BenJeau commented:
Thanks for the changes! I'm looking through it right now. The changes look good, I will test it a bit before and if there's any changes I see that we would need to do, I will push them on the branch since I want this to be merged soon so it doesn't get forgotten 😅 

I see the lockfile change, nice!

BenJeau committed:
chore: fix failing workflows

BenJeau committed:
chore: address clippy lints

BenJeau committed:
chore: address cargo fmt

BenJeau committed:
chore: update direct deps

BenJeau committed:
chore: update sub dependency to address vulnerability

BenJeau committed:
chore: fix e2e tests, increase timeouts

BenJeau committed:
chore: update README to reflect the new lockfile change

BenJeau committed:
chore: fix cargo audit not installing in GH actions

BenJeau committed:
chore: force cargo install in CI workflows

BenJeau commented:
LGTM! For the future, I'm more likely to merge changes faster if the pull request is smaller, e.g. this could have been split in multiple merge requests, just because it is easier to tests/review.

Regardless, thanks a lot for the bug fixes and the new features! 🥇
